### PR TITLE
Fix to #30028 - Added nullable property to Json mapped model resulting in errors instead of mapping non existing json property to null

### LINKED ISSUE #30028: Added nullable property to Json mapped model resulting in errors instead of mapping non existing json property to null
## File a bug

Please see #29219 for context. This issue is precisely described in the comment https://github.com/dotnet/efcore/issues/29219#issuecomment-1356886149 that was added and is related to #29219.

When a new nullable property is added to a model mapped as json, then querying on that should not error, but instead missing json values in the data should just map to null. According to #29219 looks like this was fixed for simple types (string, int, ..) but still does not work for complex types.

### Include provider and version information

EF Core version: 7.0.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 7.0

### LINKED ISSUE #30062: KeyNotFoundException on nullable nested object in JSON column if that object does not exist.
Possible duplicate of [#30028](https://github.com/dotnet/efcore/issues/30028).

EF Core 7.0 will throw a KeyNotFoundException if a JSON column does not have a nullable sub-object in the record. Code sample can be found at [EFCore7JsonDemo](https://github.com/thomasjbarrett82/EFCore7JsonDemo). This error will not happen if the classes are mapped to SQL tables instead of JSON.

## Screenshots

get all error

get one error

## Repro steps: 

1. Add a new PersonJson without an Education property - {"Id":"3fa85f64-5717-4562-b3fc-2c963f66afa6","Name":"string"}
2. Run a debug session with the Education property on the PersonJson object and try a Get or GetAll via Swagger

EF Core version: 7.0.2
Database provider: Microsoft.EntityFrameworkCore.SqlServer (SQL Server 2019)
Target framework: NET 6.0
Operating system: Windows 11
IDE: Visual Studio 2022 17.4

### PR DESCRIPTION
**Pull request by maumar**
Problem was that we when accessing inner property on a JsonElement we used GetProperty. If property is not present (which should be allowed if we try to access optional navigation) KeyNotFound is thrown.

Fix is to use TryGetProperty instead to gracefully handle this scenario.

Fixes #30028

### CODE REVIEW

**Comment by ajcvickers** [comment]
@roji Can you take a look at this? We're considering it for a patch.

File: test/EFCore.Relational.Specification.Tests/Query/JsonQueryAdHocTestBase.cs [hunk 1]
```diff
@@ -88,6 +88,135 @@ public class MyJsonEntity29219
 
     #endregion
 
+    #region 30062
+
+    protected abstract void Seed30062(MyContext30062 ctx);
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/JsonQueryAdHocTestBase.cs:93]: Should it be 30028 throughout?

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs [hunk 2]
```diff
@@ -1154,16 +1154,34 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
...
                 var innerJsonElementAssignment = Expression.Assign(
                     innerJsonElementParameter,
-                    Expression.Convert(
-                        Expression.Call(
-                            jsonElementShaperLambdaParameter,
-                            JsonElementGetPropertyMethod,
-                            Expression.Constant(ownedNavigation.TargetEntityType.GetJsonPropertyName())),
-                        typeof(JsonElement?)));
+                    Expression.Condition(
+                        Expression.Equal(
```
Comment (roji) [inline src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:1176]: nit: should be able to remove the equality-to-true no?

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs [hunk 3]
```diff
@@ -1398,12 +1416,47 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
...
+                        _variables.Add(resultFoundParameter);
+                        _variables.Add(tempParameter);
+
+                        var tryGetPropertyCall = Expression.Call(
                             Expression.MakeMemberAccess(
                                 currentJsonElementVariable!,
                                 _nullableJsonElementValuePropertyInfo),
-                            JsonElementGetPropertyMethod,
-                            Expression.Constant(stringPath));
+                            JsonElementTryGetPropertyMethod,
```
Comment (roji) [inline src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:1432]: nit: rename the static readonly PropertyInfo/MethodInfos to PascalCase (e.g. _nullableJsonElementValuePropertyInfo -> NullableJsonElementValuePropertyInfo)

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs [hunk 4]
```diff
@@ -1398,12 +1416,47 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
                     Expression jsonElementAccessExpressionFragment;
                     if (currentPath.JsonPropertyName is string stringPath)
                     {
-                        jsonElementAccessExpressionFragment = Expression.Call(
+                        // JsonElement? jsonElement = (...) <- this is the previous one
+                        // JsonElement temp;
+                        // bool resultFound = jsonElement.HasValue ? jsonElement.Value.TryGetProperty("PropertyName", temp) : false;
+                        // JsonElement? newJsonElement = resultFound == true ? (JsonElement?)temp : null;
```
Comment (roji) [inline src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:1422]: Maybe possible to simplify to:

```c#
JsonElement? newJsonElement = jsonElement.HasValue && jsonElement.Value.TryGetProperty("PropertyName", temp) ? (JsonElement?)temp : null;
```

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs [hunk 5]
```diff
@@ -1461,26 +1514,26 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
...
-                            currentJsonElementVariable!.Type),
-                        Expression.Default(currentJsonElementVariable!.Type));
+                        var jsonElementValueExpression = Expression.Condition(
+                            Expression.MakeMemberAccess(
+                                currentJsonElementVariable!,
+                                _nullableJsonElementHasValuePropertyInfo),
+                            Expression.Convert(
+                                jsonElementAccessExpressionFragment,
+                                currentJsonElementVariable!.Type),
+                            Expression.Default(currentJsonElementVariable!.Type));
```
Comment (roji) [inline src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:1525]: Bang unnecessary

**Review by roji — APPROVED** [review]
Some possible simplifications but otherwise LGTM

File: test/EFCore.Relational.Specification.Tests/Query/JsonQueryAdHocTestBase.cs (hunk 1 above)
Comment (maumar) [reply test/EFCore.Relational.Specification.Tests/Query/JsonQueryAdHocTestBase.cs:93]: 30062 is one of the duplicates, must have written the tests before we collapsed them or something. will fix

File: src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs (hunk 4 above)
Comment (maumar) [reply src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:1422]: will give it a try

**Pull request merged**